agent.go

```diff
@@ -787,6 +795,85 @@ func (a *Agent) addRemotePassiveTCPCandidate(remoteCandidate Candidate) {
 		a.localCandidates[localCandidate.NetworkType()] = append(a.localCandidates[localCandidate.NetworkType()], localCandidate)
 		a.candidateNotifier.EnqueueCandidate(localCandidate)
 
+		a.log.Debugf("Adding active tcp host candidate. LocalCandidate: %v RemoteCandidate: %v",
+			localCandidate.addr(), remoteCandidate.addr())
+		a.addPair(localCandidate, remoteCandidate)
+	}
+
+	// Loop through the list of STUN servers to create local active tcp server reflexive candidates
+	// to pair with the remote passive tcp candidate. To keep the list of generated candidates
+	// small, we only make queries until we get a single successful response. We expect further
+	// responses to discover the same external IP address.
+	for _, stunUri := range a.urls {
+		// The `stunUri` can be a STUN or TURN uri. We assume both will respond to `BindRequest`s.
+		client, err := stun.DialURI(stunUri, &stun.DialConfig{})
+		if err != nil {
+			a.log.Infof("dial err: %v", err)
+			continue
+		}
+
+		bindRequest, err := stun.Build(stun.BindingRequest, stun.TransactionID)
+		if err != nil {
+			a.log.Warnf("Error building bind request: %v", err)
+			continue
+		}
+
+		var srflxAddr net.IP
+		if err := client.Do(bindRequest, func(ev stun.Event) {
+			if ev.Error != nil {
+				a.log.Warnf("TCPActive ServerReflexive BindRequest error: %v", ev.Error)
+				return
+			}
+
+			var addr stun.XORMappedAddress
+			if err = addr.GetFrom(ev.Message); err != nil {
+				a.log.Warnf("BindResponse did not have XORMappedAddress: %v", err) //nolint:errorlint
+				return
+			}
+
+			srflxAddr = addr.IP
+		}); err != nil || srflxAddr == nil {
+			a.log.Warnf("TCPActive ServerReflexive BindRequest error/malformed output. Err: %v Addr: %v",
+				err, srflxAddr)
+			continue
+		}
+
+		// Create a connection to the passive tcp remote candidate. This async dials in a loop. This
+		// is necessary in case the remote candidate is a relay candidate. And the other peer must
+		// learn of our IP such that it can invoke `CreatePermissions` on the relay with our
+		// external IP.
+		conn := newActiveTCPConn(
+			a.context(),
+			// Bind on any available local port.
+			net.JoinHostPort(srflxAddr.String(), "0"),
```

Comment:
Ah, this is why closing -> redialing was working. I had changed this (the caller) to pass in port 0. Such that each attempt would get an available port. Without needing to muck with the dialer's localaddr.

Reply:
Before this, `newActiveTCPConn` was being called with, e.g: `127.0.0.1:55439`.

Thus creating a dialer:
```
dialer := net.Dialer{
  LocalAddr: `127.0.0.1:55439`
}
```

Closing and redialing that would get the "port in use" which seems incorrect, but we've both observed previously. Hence the error handling of:
```
if errWritingToConn != nil {
  conn.Close()
  dialer.LocalAddr.Port = 0
  conn = dialer.Dial(...)
}
```

But when I retested whether that `Port = 0` was necessary (or maybe that evolved to `dialer.LocalAddr = nil`), and saw it was unnecessary was because now the dialer was:
```
dialer := net.DIaler{
  LocalAddr: `127.0.0.1:0`
}
```

Reply:
_Got_ it. Cool thanks.